CONTRIBUTING.md

```diff
@@ -0,0 +1,106 @@
+# Contributing
+
+Pull requests are welcome!
+
+## Getting started
+
+First [fork this repo](https://help.github.com/articles/fork-a-repo/), and
+then clone it & set the upstream repo:
+
+```bash
+git clone git@github.com:your-github-username/halyard.git
+
+   # in my case (username is lwander), the above command looks like:
+   # git clone git@github.com:lwander/halyard.git
+
+cd halyard
+
+git remote add-url upstream git@github.com:spinnaker/halyard.git
+
+  # the above command allows you to push patches to the upstream repo for 
+  # approval
+```
+
+## Configuring IntelliJ
+
+In the project's root directory run
+
+```bash
+./gradlew idea
+```
+
+and then point IntelliJ at the generated project. You shouldn't need to import
+this project, simply navigating to __File__ > __Open__ > _the project's root
+directory_ should be enough.
+
+## Running the daemon
+
+In the project's root directory run
+
+```bash
+./gradlew
+```
+
+and the daemon will build & run. The first time you do this it will take a few
+minutes as it fetches the dependencies.
+
+## Running the CLI
+
+In the project's root directory run
+
+```bash
+cd halyard-cli/
+
+make
+
+./hal --help
+```
+
+> WARNING! Since this project uses one gradle project for two executables, you
+> currently can't run `make` and `./gradlew` at the same time (one of the two
+> will likley hang indefinitely).
+
+## Project structure
+
+This is a brief overview of how the project is structured.
+
+### halyard-cli/
+
+All CLI related code is here. 
+
+ * `command.v1` defines the command structure and entrypoints into command 
+ processing.  For example, the behavior of `hal config` is defined in the
+ `ConfigCommand.java` class.
+ 
+ * `ui.v1` defines UI output behavior & formatting.
+
+### halyard-config/
+
+The config validation & structure is defined here.
+
+ * `config.v1` defines utilities for loading the halconfig.
+
+ * `errors.v1` defines error & exception classes to be thrown by the
+   validation.
+
+ * `model.v1` defines the structure of the halconfig. The `providers`
+   sub-package is important if you want to add support for a specific
+   provider's configuration options
+ 
+ * `services.v1` defines services for retrieving different parts of your
+   halconfig. These are ultimately consumed by the web controllers.
+
+ * `validate.v1` defines the validation behavior for our config. The
+   classes described in the `providers` sub-package are used in annotations for
+   specific fields/classes to allow only the necessary validators to run when
+   required.
+
+### halyard-web/
+
+The web controllers & endpoints are defined here.
+
+ * `controllers.v1` defines the set of endpoints the daemon is reachable from.
+
+ * `errors.v1` catches validation exceptions and forwards them to the client as
```

Comment:
what is a client? The CLI (talks to the web server, indicating this is probably embedded in the daemon) and/or a browser? If it is a browser then "forwarding" sounds odd as compared to "responds". But why isnt the controller doing that? This feels like a "goto".

Ok, I cheated and looked at the implementation. I guess exception handlers makes sense. They look like they are spring magic stuff so have no idea how control works but presumably the controller is implicitly delegating to these (via an unarticulated assumption of someone else introducing them to spring, which is one of my naive annoyances with spring).

If the above is right, I'd be more explicit and state that these are exception handlers for the different types of exceptions that can be thrown within a controller flow, and handle the exception by terminating the flow and responding to the request as appropriate.